### ISSUE TITLE
Almost all predictions are persons (the first class)

### ISSUE THREAD

**Issue by realwecan** [CharlesShang/FastMaskRCNN#97]
I tried prediction with the trained model, however it seems the output are strongly biased towards the first non-background class (person):

![0454750_dets](https://user-images.githubusercontent.com/4241989/27583294-78e6263e-5b66-11e7-83b8-9683edeb85d1.png)
![0455085_dets](https://user-images.githubusercontent.com/4241989/27583317-87541a50-5b66-11e7-886f-1af0ad30f6f2.png)
![0459500_dets](https://user-images.githubusercontent.com/4241989/27583337-9b2cdc10-5b66-11e7-88d4-6f361deed4d2.png)

The above predictions are obtained from a model trained with 50K iterations.

Other class predictions are not impossible (see the 3rd image, there are a few cars), but it just doesn't seem right to have such a big bias towards person.

Could anyone suggest what shall I be looking at if I wanted to debug this issue? Thanks!

**Comment by souryuu** [comment]
I am not sure about how the original maskrcnn create the target rois for training. From the current version of code, it seems that rois from RPN will be match with the largest IOU ground truth object. However, when the ground truth objects stack on each others, rois is still matched only to the largest IOU object.

**Comment by Sharathnasa** [comment]
@tou wang did you wrote and ranned the test script to get this kind of
output?

**Comment by realwecan** [comment]
@souryuu Sorry I may have missed your point -- will this ground-truth overlap issue lead to a bias towards person?

**Comment by vidhi09** [comment]
@realwecan 
Hello,
Culd you just let me know about the code that you have used for testing the data.Even I have trained the model.But not understanding how to generate the test code.Hope you would consider my request

**Comment by souryuu** [comment]
@realwecan For me it seems that large portion of the images contain person in it (both in term of size and frequency). Also after I trained it up to some point, it seems bias toward not only to person but some particular large and frequently appearing objects.

**Comment by realwecan** [comment]
@souryuu This is strange and could be problematic, as we would like to get rid of this bias if it indeed exists. For how many iterations have you trained your model with so far? I would appreciate if you could share your trained model with me, and I can do some detailed evaluation.

**Comment by HuangBo-Terraloupe** [comment]
Can you please share the code of evaluation and trained weights?

**Comment by souryuu** [comment]
@realwecan I am about to get 700k tmr (total training time = 5days). I did not focus on the mask part yet. I saw some of them on tensorboard but they did not look good.

**Comment by vidhi09** [comment]
@realwecan  @souryuu 
Hello,
This is Nishith.I am Btech student studying in Indian Institute of technology Varanasi.Being new to field of deep learning I would like to request you to provide some direction in the field of testing as the testing code is not provided.
Awaiting your reply

**Comment by zhuwenzhen** [comment]
@souryuu could you share your 700k trained weights?

**Comment by souryuu** [comment]
@zhuwenzhen I can not share the weight because of the restriction in my working environment.
@vidhi09 I do not have the code for testing the network. However if you want to look at the output of bounding box regression and classification during training, you can follow some instruction in #96. 
@realwecan Could you please share the testing code? It would be nice if we can check segmentation of all bounding boxes at once like in your result.

**Comment by realwecan** [comment]
@souryuu Acutally, I obtained the segmentation masks with your training codes (on validation data, without updating network weights). I will work on a separate testing codes and then share with everyone.

**Comment by Nishith365** [comment]
@realwecan 
could you just elaborate on your idea of visualizing the masks.So I have to give a demo next week with the masks.So If you could provide me the idea with the visualizing of the masks that would be help full.

**Comment by Nishith365** [comment]
@souryuu 
In your issue #96 you have mentioned that "Removing the comment between line 284-291 in train.py should save the image shown above in output/maskrcnn/est_imgs/"  but train.py has only around 219 lines.
Could you just elaborate on your method.being the beginner i am finding it difficult to understand.Hope you would consider my request.

**Comment by Duankaiwwen** [comment]
@Nishith365    https://github.com/souryuu/FastMaskRCNN

**Comment by Nishith365** [comment]
@souryuu 
Thank you

**Comment by brisker** [comment]
@realwecan 
@souryuu 
@CharlesShang 
In the original paper, "the mask target is the intersection between an RoI and its associated ground-truth mask." . But it seems that the code here: https://github.com/CharlesShang/FastMaskRCNN/blob/master/libs/layers/mask.py#L73-L76
 just use the associated ground-truth mask, not the intersection between an RoI and its associated ground-truth mask. So is there any bug in the code , or have I understand something wrong?
